tests/testsuite/publish.rs

```diff
@@ -2048,3 +2049,187 @@ error: package ID specification `bar` did not match any packages
+[UPDATING] `dummy-registry` index
+",
+        )
+        .run();
+
+    // Verify the responder has been pinged
+    let lock = arc2.lock().unwrap();
+    assert_eq!(*lock, 2);
+    drop(lock);
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+                [dependencies]
+                delay = "0.0.1"
+            "#,
+        )
+        .file("src/main.rs", "fn main() {}")
+        .build();
+
+    p.cargo("build -Z sparse-registry")
+        .masquerade_as_nightly_cargo(&["sparse-registry"])
+        .with_status(0)
+        .run();
+}
+
+/// A separate test is needed for package names with - or _ as they hit
+/// the responder twice per cargo invocation. If that ever gets changed
+/// this test will need to be changed accordingly.
+#[cargo_test]
+fn wait_for_publish_underscore() {
+    // Counter for number of tries before the package is "published"
+    let arc: Arc<Mutex<u32>> = Arc::new(Mutex::new(0));
+    let arc2 = arc.clone();
+
+    // Registry returns an invalid response.
+    let _registry = registry::RegistryBuilder::new()
+        .http_index()
+        .http_api()
+        .add_responder("/index/de/la/delay_with_underscore", move |req, server| {
+            let mut lock = arc.lock().unwrap();
+            *lock += 1;
+            // package names with - or _ hit the responder twice per cargo invocation
+            if *lock <= 2 {
+                server.not_found(req)
+            } else {
+                server.index(req)
+            }
+        })
+        .build();
+
+    // The sparse-registry test server does not know how to publish on its own.
+    // So let us call publish for it.
+    Package::new("delay_with_underscore", "0.0.1")
+        .file("src/lib.rs", "")
+        .publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "delay_with_underscore"
+                version = "0.0.1"
+                authors = []
+                license = "MIT"
+                description = "foo"
+
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("publish --no-verify -Z sparse-registry")
+        .masquerade_as_nightly_cargo(&["sparse-registry"])
+        .with_status(0)
+        .with_stderr(
+            "\
+[UPDATING] `dummy-registry` index
+[WARNING] using `registry.token` config value with source replacement is deprecated
+This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/xxx>.
+Use the --token command-line flag to remove this warning.
+[WARNING] manifest has no documentation, [..]
+See [..]
+[PACKAGING] delay_with_underscore v0.0.1 ([CWD])
+[UPLOADING] delay_with_underscore v0.0.1 ([CWD])
+[UPDATING] `dummy-registry` index
+[WAITING] on registry `dummy-registry` to finish publish
```

Comment:
Calling out the waiting message since its likely people will have feedback on it

Current behavior:
- If the publish is complete before the wait, no waiting message will be shown
- If the publish isn't complete, we'll show the message immediately but only once

I was tempted to put the message on a time delay but
- Since we show the "Updating" messages, I wanted there to be context for them (I was tempted to put the "Waiting" before the first message because of this)
- It'd make it more annoying to test

Reply:
This is now
```
[WAITING] on `delay_with_underscore` to be in `dummy-registry` index
```
Because we switched from `PackageRegistry` to a `Source`, the desciption changed from `registry 'dummy-registry'` to `'dummy-registery' index`, so I re-worded to better fit that.

Reply:
Maybe a minor tweak
```
[WAITING] on `delay_with_underscore` to propagate to `dummy-registry` index
```